microsoft/fluent-ui-react
# chore: add Popup issue-example-based e2e tests

This PR makes browser tests to execute against encapsulated React examples. 

With these changes it will be possible to easily introduce coverage for those problems that are reported in issues, given that issue contains a repro example - and, by that, ensure that each previously problematic scenario remains to be a working one after each new change to the code.

## Commands (yarn scripts)
- `test:e2e` (`yarn test:e2e`) - independent task: builds, serves and runs tests
- `test:e2e:serve` (`yarn test:e2e:serve`) - builds and serves site with E2E examples. Index page of this site contains link to all E2E examples

## HOW TO - add browser test with custom example
Lets suppose that one would like to introduce tests for some scenario with `<scenarioName>`. This is set of steps to follow:
- all test files should be added  to `e2e/tests` directory of the repo
- create React example file named as `<scenarioName>-example.tsx`
  - provide component to render, export it as the default one
  - to check how it renders, run `yarn test:e2e:serve` and go to `/scenario-name` URL (kebab-cased name of scenario) - or go to the index page (`/`) and search for the link to your example there.
- create test file named `<scenarioName>-test.ts`
  - to render corresponding example, use `await e2e.gotoTestCase('<scenarioName>')`. This call is usually made in `beforeEach()` test setup function
  - for safe rename refactoring, it is recommended to use `__filename` variable of NodeJS as an argument:
     - `e2e.gotoTestCase(__filename)`
     - this `gotoTestCase` function will trim `-(example|test).tsx?` postfix from file name, to leave only `<scenarioName>` part

## EXAMPLE - Popup tests
As a working example, this PR contains couple of `Popup` tests for previously discovered and addressed issues:
- https://github.com/stardust-ui/react/issues/557 (menu item's popup is not closing when another menu item is clicked)
- https://github.com/stardust-ui/react/issues/1079 (incorrect handling of ESC key)
- https://github.com/stardust-ui/react/issues/1324 (incorrect click handling when popup content is clicked)
  - ~~is disabled for now, as this test case doesn't pass~~
  - ~~will be enabled with #1482, where fix is introduced~~
